apps/emqx/src/emqx_persistent_session_ds.erl

```diff
@@ -465,6 +472,12 @@ handle_timeout(
         end,
     Session = emqx_session:ensure_timer(?TIMER_PULL, Timeout, Session1),
     {ok, Publishes, Session};
+handle_timeout(ClientInfo, ?TIMER_PULL, Session0 = #{replay := [_ | _]}) ->
+    Session = replay_streams(Session0, ClientInfo),
+    {ok, [], Session};
+handle_timeout(ClientInfo, ?TIMER_RETRY_REPLAY, Session0) ->
+    Session = replay_streams(Session0, ClientInfo),
+    {ok, [], Session};
```

Comment:
This part looks wrong: identical code runs for two different timer events. 
We reset timers manually, and as such we must maintain the invariant: handler of every timer event `X` MUST end with resetting timer `X`. This way we can make sure that we never break the loop. 

To keep the flow simple, I suggest keeping `replay_stream` out of TIMER_PULL loop. It may check if replay is ongoing, but not do the replay.